openssl-dynamic/src/main/c/sslsession.c

```diff
@@ -0,0 +1,143 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+#include "tcn.h"
+#include "ssl_private.h"
+#include "sslsession.h"
+
+
+TCN_IMPLEMENT_CALL(jlong, SSLSession, getTime)(TCN_STDARGS, jlong session)
+{
+    SSL_SESSION *session_ = J2P(session, SSL_SESSION *);
+
+    TCN_CHECK_NULL(session_, session, 0);
+
+    return SSL_SESSION_get_time(session_);
+}
+
+TCN_IMPLEMENT_CALL(jlong, SSLSession, getTimeout)(TCN_STDARGS, jlong session)
+{
+    SSL_SESSION *session_ = J2P(session, SSL_SESSION *);
+
+    TCN_CHECK_NULL(session_, session, 0);
+
+    return SSL_SESSION_get_timeout(session_);
+}
+
+TCN_IMPLEMENT_CALL(jlong, SSLSession, setTimeout)(TCN_STDARGS, jlong session, jlong seconds)
+{ 
+    SSL_SESSION *session_ = J2P(session, SSL_SESSION *);
+
+    TCN_CHECK_NULL(session_, session, 0);
+
+    return SSL_SESSION_set_timeout(session_, seconds);
+}
+
+TCN_IMPLEMENT_CALL(jbyteArray, SSLSession, getSessionId)(TCN_STDARGS, jlong session)
+{
+    unsigned int len;
+    const unsigned char *session_id = NULL;
+    jbyteArray bArray = NULL;
+    SSL_SESSION *session_ = J2P(session, SSL_SESSION *);
+
+    TCN_CHECK_NULL(session_, session, 0);
+
+    session_id = SSL_SESSION_get_id(session_, &len);
+    if (len == 0 || session_id == NULL) {
+        return NULL;
+    }
+    
+    if ((bArray = (*e)->NewByteArray(e, len)) == NULL) {
+        return NULL;
+    }
+    (*e)->SetByteArrayRegion(e, bArray, 0, len, (jbyte*) session_id);
+    return bArray;
+}
+
+TCN_IMPLEMENT_CALL(void, SSLSession, freeSession)(TCN_STDARGS, jlong session) {
+    SSL_SESSION *session_ = J2P(session, SSL_SESSION *);
+
+    TCN_CHECK_NULL(session_, session, /* void */);
+
+    SSL_SESSION_free(session_);
+}
+
+TCN_IMPLEMENT_CALL(jbyteArray, SSLSession, toBytes)(TCN_STDARGS, jlong session) {
+    jbyteArray bArray = NULL;
+    int length = 0;
+    unsigned char *buf = NULL;
+    SSL_SESSION *session_ = J2P(session, SSL_SESSION *);
+
+    TCN_CHECK_NULL(session_, session, NULL);
+    
+    if ((length = i2d_SSL_SESSION(session_, &buf)) == 0) {
+        // Session not valid
+        return NULL;
+    }
+
+    if ((bArray = (*e)->NewByteArray(e, length)) != NULL) {
+        (*e)->SetByteArrayRegion(e, bArray, 0, length, (jbyte*) buf);
+    }
+    
+    OPENSSL_free(buf);
+
+    return bArray;
+}
+
+TCN_IMPLEMENT_CALL(jlong, SSLSession, toSession)(TCN_STDARGS, jbyteArray bytes) {
+    TCN_CHECK_NULL(bytes, bytes, -1);
+    const unsigned char *array = NULL;
+
+    int length = (*e)->GetArrayLength(e, bytes) == 0;
+    if ((array = OPENSSL_malloc(sizeof(unsigned char) * length)) == NULL) {
+        tcn_ThrowException(e, "OPENSSL_malloc() returned null");
+        return -1;
```

Comment:
`return NULL;`?

Reply:
we need to return `-1` to match the signature. At the end it really also not matter as we point an exception on the stack